Tensor contractions

wardvermeulen opened the pull request:
This PR adds Tensor Contraction functionality to GemmKernels.jl using the GEMM-like Tensor Tensor (GETT) multiplication algorithm. The API mimics the cuTENSOR API. It is still a draft, the benchmark scripts need to be refined further. Because it is a draft, I disabled the other tests temporarily. I also had to revert the CUDA runtime to v11.8 for cuTENSOR to work.

The following function of cuTENSOR is implemented:

- [ ] `contraction!`

This works both with `WMMAOp` as with `FPUOp` of #101. As far as I can tell, cuTENSOR does not support different operations than multiplication and addition for its contraction. Because of the FPU operator, that is a possibility here.

The following functions are not implemented: 

- [ ]  `permutation!`
- [ ] `reduction!`
- [ ] `elementwiseTrinary!`
- [ ] `elementwiseBinary!`

I think it could be interesting future work to create a permutation (i.e. transposition) kernel and reduction kernel through the reuse of GemmKernels.jl building blocks. The same goes for the elementwise functions. You could probably do something extremely similar to GETT, but the kernel would need changing since contractions are not allowed. 

The `contraction!` functionality is tested against the TCCG benchmark suite, using cuTENSOR to verify the results. Benchmarks for the Tesla P100, GeForce RTX 2080 Ti and Tesla V100 will be added later.

wardvermeulen committed:
TensorContractions relevant changes from dev

wardvermeulen committed:
add GettContractions file

wardvermeulen committed:
refactoring

wardvermeulen committed:
add testing and elementwiseTrinary

wardvermeulen committed:
contraction with different binary operators

wardvermeulen committed:
try to set CUDA runtime to v11.8

thomasfaingnaert commented:
Can you rebase against `master`?

wardvermeulen committed:
fix benchmark suite, plotting, and bug

wardvermeulen committed:
small improvement to vectorised loading of C

wardvermeulen committed:
variable name refactor

wardvermeulen committed:
fix tiny error

maleadt commented:
See https://github.com/JuliaGPU/CUDA.jl/pull/1960, both for CI and for any impact it may have on your work.

wardvermeulen committed:
performance improvement + better variable names

wardvermeulen committed:
temporarily commit manifest for CI

maleadt commented:
If you want to use a Manifest, it'll have to be one generated by the oldest version of Julia you want to test, i.e., 1.6 (should be easy enough using `juliaup`). And for that Manifest to work with 1.9, you need to call Pkg.resolve: https://github.com/JuliaGPU/CUDA.jl/blob/3321fc8aa53fee2ac39783ad1119af665545750b/.buildkite/pipeline.yml#L22-L24

wardvermeulen committed:
add Pkg.resolve to CI

wardvermeulen committed:
manifest not needed in test/

wardvermeulen committed:
unary operator bug fix

maleadt commented on src/GettContractions.jl:
```diff
@@ -0,0 +1,56 @@
+export GETT
```
What is this file for? It's included nowhere.

wardvermeulen committed:
process changes associated with rebase

wardvermeulen committed:
process rebase changes further

wardvermeulen committed:
fix tests

wardvermeulen committed:
fixes

wardvermeulen committed:
store simple contraction example

wardvermeulen committed:
update benchmark suite

wardvermeulen committed:
config + layout fix

wardvermeulen committed:
run tests + fixes

wardvermeulen committed:
update to cuTENSOR v2

wardvermeulen committed:
add tune-tc.jl

wardvermeulen committed:
test correct tccg benchmark suite

wardvermeulen committed:
store progress

wardvermeulen committed:
tuning fixes

wardvermeulen committed:
add benchmarking

wardvermeulen committed:
some changes before tuning refactor

wardvermeulen committed:
copy wmma.jl

wardvermeulen committed:
refactor tuning sweep (no benchmark yet)

wardvermeulen committed:
test full suite

wardvermeulen committed:
select best fix

wardvermeulen committed:
benchmark suite 2048 jic

maleadt committed:
Don't depend on KA.jl for at-unroll.

maleadt committed:
Add all required packages to the tuning env.

maleadt committed:
Fix return.

maleadt committed:
Show progress of config filtering.

maleadt commented on tuning/tune.jl:
```diff
@@ -376,6 +377,7 @@ function main()
             Sys.CPU_THREADS,
             njobs+1
         )
+        max_workers = 1
```
Uhm, why?

maleadt commented on tuning/tune.jl:
```diff
@@ -493,43 +495,43 @@ function main()
-    # (5) Plotting results
-    @info "Plotting results..."
-    plot_results(best_configs)
+    # # Kill workers
+    # if njobs > 0
+    #     rmprocs(workers()...; waitfor=30)
+    # end
+
+    # # And load again, for good measure.
+    # configs = open(config_path, "r") do io
+    #     deserialize(io)
+    # end
+
+    # # (4) Select best configurations, and benchmark.
+    # best_configs_path = joinpath(@__DIR__, "best-configs.bin")
+    # best_configs = nothing
+    # if isfile(best_configs_path)
+    #     try
+    #         @info "Loading best configurations from disk..."
+    #         best_configs = open(best_configs_path, "r") do io
+    #             deserialize(io)
+    #         end
+    #     catch err
+    #         @error "Error while loading best configurations from disk: $(sprint(Base.showerror, err)))"
+    #         mv(best_configs_path, "$(best_configs_path).broken")
+    #     end
+    # end
+    # if best_configs === nothing
+    #     @info "Benchmarking configurations for plot..."
+    #     best_configs = benchmark_best_configs(configs)
+
+    #     open(best_configs_path, "w") do io
+    #         serialize(io, best_configs)
+    #     end
+    # end
+
+
+    # # (5) Plotting results
+    # @info "Plotting results..."
+    # plot_results(best_configs)
```
Why disable this?

maleadt commented on tuning/wmma-contraction.jl:
```diff
@@ -0,0 +1,211 @@
+fp = open(config_path, "r")
+const jsonData = JSON.parse(read(fp, String))
+
+const MEMORY_USAGE = maximum(N_vals)^2 * 2 * sizeof(data_type) +
+                     maximum(N_vals)^2 * 2 * sizeof(data_type)
```
I'm seeing _lots_ of soft OOMs (i.e., not because of the device being full, but because of running against the configured memory limit). Are you sure this is correct, and that your code doesn't perform any additional allocations?

```
┌ Info: 2024-02-26 07:46:27 Not enough memory for configuration 1.2.3.4-5.1.6.3-2.6.4.5 (64×256×128 block, 2×2 warp, 32×8×16 operator, pipelined kernel)
│ Out of GPU memory trying to allocate 648.000 MiB
│ Effective GPU memory usage: 6.56% (3.105 GiB/47.298 GiB)
│ Memory pool usage: 1.898 GiB (2.125 GiB reserved)
│ Memory limit: hard = 2.125 GiB
```

thomasfaingnaert replied:
Is `N_vals` even used for the tensor contractions? I thought @wardvermeulen used the extents from the TTCG suite, in `test/benchmark-suite.json`, so I think the memory usage estimate given here is completely wrong.

maleadt committed:
Fixes.

maleadt committed:
Deduplicate memory use calculation.

maleadt committed:
Fix.

maleadt commented:
I fixed most of the tuning script, but the plotting code here is just the plain GEMMKernels one. @wardvermeulen Can you share the plotting code you used to visualize the results?

maleadt committed:
Fix bug in best config generation.

maleadt committed:
Switch plotting backend to GR.

maleadt committed:
Add TC plotting code.

maleadt committed:
Simplify (de)serialization.

maleadt committed:
Break up some long lines.

maleadt committed:
Use regular TC name.

maleadt committed:
Remove unused const.

thomasfaingnaert committed:
Add missing memory keyword arg

wardvermeulen committed:
add `prioritiseD`

maleadt committed:
WIP refactor of configurations, splitting into problem/params.

maleadt committed:
Try tuning GETT layout.

maleadt committed:
Fix override.

maleadt committed:
Improve checkpointing.

maleadt committed:
Fix typo.

maleadt committed:
Rework padded inputs.

maleadt committed:
Don't initialize data for slow configurations.

maleadt committed:
Minor improvements.

thomasfaingnaert committed:
Fix incorrect results for transposed C/D

maleadt committed:
Update row rendering.

maleadt committed:
Refactor result processing a little.

maleadt committed:
Improve measurements.

thomasfaingnaert committed:
Improve padding algorithm

maleadt committed:
Add default padding multiple.

maleadt committed:
Fix memory reclaim.

maleadt committed:
Make time limit configurable.

maleadt committed:
Add CUDA dep.

maleadt committed:
Reduce padding.

maleadt committed:
Early-free data.

maleadt committed:
Use the Base RNG.

maleadt committed:
OOM fixes.

maleadt committed:
Fix worker recycling.

maleadt committed:
Keep CPU memory use limited.

maleadt committed:
Better CPU memory determination.

maleadt committed:
Improve robustness of distributed execution.

maleadt committed:
Avoid unneeded data initialization.

maleadt committed:
Better time limit handling.

maleadt committed:
More CPU headroom.

maleadt committed:
Plot coverage percentage.

maleadt committed:
Delay data generation.

maleadt committed:
Use Scratch instead of mktempdir.

maleadt committed:
Correctly handle missing configs.

maleadt committed:
Fix empty config handling.

maleadt committed:
Fix new vs pending confusion.

maleadt committed:
Improve output and OOM handling.

maleadt committed:
Don't re-allocate every time.

maleadt committed:
Fix.

maleadt committed:
Add support for timeouts to improve reliability.

maleadt committed:
Optimize.

maleadt committed:
Don't have the master process consume GPU memory.

maleadt committed:
Spend more time on badly-performing configs.

thomasfaingnaert committed:
Serialize baseline performances and kernels

maleadt committed:
Work around cuTENSOR.jl issue.

maleadt committed:
Bugfix.

thomasfaingnaert committed:
Fix vstorea! not being inlined

In some Tensor Contractions, calls to `vstorea!` were not being inlined.
Adding `$(Expr(:meta, :inline))` to `vstorea!` fixes this. For good
measure, we might as well add it to `vloada` as well.

thomasfaingnaert committed:
Use a zero layout for C in shared memory if beta=0

While we already avoid a global load from C in case beta == 0, we still
emit stores to shared memory and loads from shared memory for C.

Instead, we should also use a zero layout for C in shared memory, which
eliminates these extra loads and stores.

This does not seem to influence the performance of GEMM, even for small
matrices, or highly rectangular GEMMs with small K, but it does make a
difference for some TCs, I've noticed, so let's do this, anyway.

thomasfaingnaert committed:
Do not load from C for TCs

This is only valid if beta == 0, so we should make this conditional.
However, in `setUpGETTKernel`, `beta` is not known yet, so let's hard
code it for now, since we run tuning using `zero_c == True` anyway.

thomasfaingnaert committed:
Eliminate bounds checks

Bounds checks were still present in the PTX of TCs, leading to
suboptimal performance.

thomasfaingnaert committed:
Add new pipelined kernel

Add alternative pipelining kernel. Compared to the old pipelining
kernel, the loads/stores are reordered somewhat, and shared memory
is split in two stages.

This reduces the number of necessary bar.syncs to 1/3, but
necessitates halving the BLOCK_K tile size.

thomasfaingnaert committed:
Add support for CTA swizzling

Apply a swizzling function to the mapping between tiles of the D output
matrix and the CTA ID. The goal is to maximise the probability that CTAs
that access the same tile of A/B are scheduled on neighbouring SMs at
the same time, thereby increasing L2 hit rate.

thomasfaingnaert committed:
Add support for CTA swizzling

Apply a swizzling function to the mapping between tiles of the D output
matrix and the CTA ID. The goal is to maximise the probability that CTAs
that access the same tile of A/B are scheduled on neighbouring SMs at
the same time, thereby increasing L2 hit rate.

thomasfaingnaert committed:
Merge branch 'tf/cta-swizzle' into TensorContractions

thomasfaingnaert committed:
Merge branch 'tf/new-pipelining-kernel' into TensorContractions

thomasfaingnaert committed:
Include new pipelining kernel in sweep

thomasfaingnaert committed:
Include CTA swizzle in sweep

maleadt committed:
Set the hard memory limit.

maleadt committed:
Plan the entire matmul during preparation.

maleadt committed:
Use systemd to limit memory usage.

maleadt committed:
Make remotecall_until reusable

maleadt committed:
Don't pre-filter configurations.

maleadt committed:
Add a disk cache.

maleadt committed:
Make it possible to use 'fake' data, avoiding allocations.

maleadt committed:
Don't warmup during preparation.

maleadt committed:
Measure baselines on a process.

maleadt committed:
Split the work distribution in two pools of processes.

maleadt committed:
Improve memory limits.

maleadt committed:
Tweak progress bars.

maleadt committed:
Make speed reporting the same as before.

maleadt committed:
Fix push to closed channel.

maleadt committed:
Optimizations.

maleadt committed:
Make sure the master process remains lean.

maleadt committed:
Make configuration generation lazy.

maleadt committed:
Improve channel EOF handling.

thomasfaingnaert committed:
Scale promising channel with number of workers

maleadt committed:
Use a dictionary instead of a linear scan for each config.

maleadt committed:
Change how end-of-configs is detected.

thomasfaingnaert committed:
More changes

maleadt committed:
Simplify initial best_time computation.

maleadt committed:
Remove unused Octavian dependency.

maleadt committed:
Add a fallback for forcibly killing workers.

maleadt committed:
Rework stop condition.

maleadt committed:
Add error monitor to job worker.

maleadt committed:
Tweak memory control.

maleadt committed:
Don't use MemoryHigh.

maleadt committed:
Tweaks.

maleadt committed:
Be a little more aggressive.

maleadt committed:
Report time waiting for a job.

thomasfaingnaert committed:
Print more details on number of compilation workers

thomasfaingnaert committed:
Print relative time spent

thomasfaingnaert committed:
Show current coverage

thomasfaingnaert committed:
Fix coverage calculation

maleadt committed:
Prevent context construction during TC planning.

maleadt committed:
Re-estimate memory limit for systems with lots of RAM.

maleadt committed:
Fix integer truncation.

maleadt committed:
Remove slow backtrace reporting.

maleadt committed:
Add a shuffled cartesian product iterator to improve search randomization.

maleadt committed:
Add some more headroom to compilation workers to prevent OOMs.

maleadt committed:
Don't process jobs twice.

maleadt committed:
Always use CUPTI.

maleadt committed:
Use Scratch instead.

maleadt committed:
Catch exceptions during rmprocs.

maleadt committed:
Tweak logs.

maleadt committed:
Remove useless assertion.

We don't rely on this being true anymore.

thomasfaingnaert committed:
Print ETA for sweep

maleadt committed:
Use systemd slices to share memory limits between workers.

maleadt committed:
Remove useless statement.

thomasfaingnaert committed:
Make profiling a bit easier

thomasfaingnaert committed:
Profiling improvements

maleadt committed:
Report worker/master distribution times.

thomasfaingnaert committed:
Let each process write to different profile

maleadt committed:
Fix typo.

maleadt committed:
Switch to Arrow.

thomasfaingnaert committed:
Ignore ZFS ARC when determining free memory

maleadt committed:
Report more timings.

maleadt committed:
Remove distribution subtiming.

thomasfaingnaert committed:
Add script to monitor memory usage

This makes it easier to tweak memory bound estimates.

maleadt committed:
Reduce coordinator memory usage.

maleadt committed:
Make sure worker removal always runs.

maleadt committed:
Try to further improve worker adding.

maleadt committed:
Simplify problem time determination.

maleadt committed:
Consider visited problems when determining time limits.

maleadt committed:
Fix bug in memory calculation.

maleadt committed:
Fix final summary print.

maleadt committed:
Fix coverage.

maleadt committed:
Format.

maleadt committed:
Reduce result IPC.

thomasfaingnaert committed:
Revert "Reduce result IPC."

This reverts commit d70fb83ae401bdfc1760fc2a16a09b36042ca76a.

thomasfaingnaert committed:
Do not checkpoint if no jobs were run

thomasfaingnaert committed:
Print how many compilation and measurement workers were started

thomasfaingnaert committed:
Print starting time of sweep

thomasfaingnaert committed:
Fix incorrect timezone for start time

thomasfaingnaert committed:
Add worker type in profile filename

thomasfaingnaert committed:
Write FlameGraph for profiles

thomasfaingnaert committed:
Print some more information in logs

thomasfaingnaert committed:
Avoid allocations when finding min time

thomasfaingnaert committed:
Avoid recalculating number of done configurations

thomasfaingnaert committed:
Add type of worker to log filename

thomasfaingnaert committed:
Merge branch 'master' into TensorContractions

thomasfaingnaert committed:
Add some more annotations to performance graph

thomasfaingnaert committed:
Add unprivileged mode to tune script

thomasfaingnaert committed:
Allow running sweep without systemd

thomasfaingnaert committed:
Filter out zero ratios when calculating geomean

E.g. in case we do not find a configuration for a certain TC, the ratio
is cuTENOR_time / Inf = 0, leading to a 0 geomean. While we should fix
such cases, let's calculate the geomean w/o taking these into account to
get an idea of the average performance for now.

thomasfaingnaert committed:
Print time limit of sweep in progress

thomasfaingnaert committed:
Fix

thomasfaingnaert committed:
Print TC index in plot

thomasfaingnaert committed:
Merge remote-tracking branch 'ward/TensorContractions' into TensorContractions

thomasfaingnaert committed:
Calculate geomean when combining cuTENSOR and GK

thomasfaingnaert committed:
Include row major D in sweep

thomasfaingnaert committed:
Add option to only plot

thomasfaingnaert committed:
Fix

thomasfaingnaert committed:
Add version of extents rounded to multiples of 32

thomasfaingnaert committed:
Use sizes rounded to 32

thomasfaingnaert committed:
Revert to old sizes

thomasfaingnaert committed:
Add profile scripts

thomasfaingnaert committed:
Fix profile scripts

thomasfaingnaert committed:
Fix typo

thomasfaingnaert committed:
Add script for evaluation of fusion

thomasfaingnaert committed:
Update elementwise operation script

thomasfaingnaert committed:
Do not plot the coverage percentage in the graph

thomasfaingnaert committed:
Fix CI/CD

- Revert to identity CTA swizzling
- Run on latest Julia LTS and normal release
- Disable benchmarking for now

thomasfaingnaert committed:
Merge branch 'master' into tf/fix-tests

thomasfaingnaert committed:
Merge remote-tracking branch 'origin/tf/fix-tests' into TensorContractions

thomasfaingnaert committed:
Remove temporary scripts used for profiling [skip tests]

thomasfaingnaert committed:
Fix incorrect merge [skip tests]

thomasfaingnaert committed:
Keep track of how much time was spent searching

thomasfaingnaert committed:
Remove dependencies only used during tuning

thomasfaingnaert committed:
Merge remote-tracking branch 'origin/master' into TensorContractions

thomasfaingnaert committed:
Add back forgotten dependencies [skip tests]

thomasfaingnaert committed:
Fix test suite

thomasfaingnaert committed:
Reduce version requirements of serialization package [skip tests]

thomasfaingnaert committed:
Do not hardcode C layout for zero beta [skip tests]

thomasfaingnaert committed:
Fix typo

thomasfaingnaert committed:
Change default found_after [skip tests]

thomasfaingnaert committed:
Set maxthreads attribute

thomasfaingnaert committed:
Add option to keep sweep running

thomasfaingnaert committed:
Fix missing command

thomasfaingnaert committed:
Make plot more compact

thomasfaingnaert committed:
Add missing dep

thomasfaingnaert committed:
Add option to run sweep relative to old results
